### PR TITLE
Rust and RISC-V

### PR DESCRIPTION
**Pull request by cecton**

### CODE REVIEW

File: content/posts/rust-and-riscv.md [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+["Rust, Arduino and Embedded Development as a Beginner: Part 1"]({{< ref "rust-and-arduino-part1" >}})
+for a more beginner friendly article.
+
+Introduction
+------------
+
+After my experience with the AVR microchip family (the same you use on Arduino
+boards) I decided to try out Rust on a RISC-V based board. I don't work in
+embedded development but most of the professional projects I saw were using
+either AVR either ARM-based architecture. As I understand the reason for that
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:16]:
```suggestion
either AVR or ARM-based architectures. As I understand the reason for that
```

File: content/posts/rust-and-riscv.md [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+is because they are kinda industry standards and you can find PCB assemblers
+who provides AVR and ARM chips more easily. RISC-V on the other hand is
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:18]:
```suggestion
who provide AVR and ARM chips more easily. RISC-V on the other hand is
```

File: content/posts/rust-and-riscv.md [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+relatively new (2010 according to
+[Wikipedia](https://en.wikipedia.org/wiki/RISC-V)) but is entirely open source
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:20]:
```suggestion
[Wikipedia](https://en.wikipedia.org/wiki/RISC-V)) and is entirely open source.
```

File: content/posts/rust-and-riscv.md [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+which doesn't require fees to use it.
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:21]:
```suggestion
Unlike other instruction sets, RISC-V has no license fees attached to it.
```

File: content/posts/rust-and-riscv.md [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+
+I had the chance to test two different
+chip producers for RISC-V: [SiFive](https://en.wikipedia.org/wiki/SiFive)
+(which actually outsource the actual manufacturing) and
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:25]:
```suggestion
(which outsource the actual manufacturing) and
```

File: content/posts/rust-and-riscv.md [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+[GigaDevice](https://en.wikipedia.org/wiki/GigaDevice). SiFive is what you will
+see on the most common development boards like the HiFive boards. In this
+article I'm using the
+[RED-V Thing Plus](https://www.sparkfun.com/products/15799) from SparkFun which
+has the same MCU than the
+[HiFive 1 Rev B](https://www.sifive.com/boards/hifive1-rev-b). One of the main
+advantage of the RED-V is again the
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:32]:
```suggestion
advantage of the RED-V is the
```

File: content/posts/rust-and-riscv.md [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+[Qwiic connector](https://www.sparkfun.com/qwiic) that I use in the previous
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:33]:
```suggestion
[Qwiic connector](https://www.sparkfun.com/qwiic) that I used in the previous
```

File: content/posts/rust-and-riscv.md [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+article ["Rust, Arduino and Embedded Development as a Beginner: Part 1"]({{< ref "rust-and-arduino-part1" >}}).
+This allows us to connect the exact same screen
+([Zio Qwiic OLED Display (1.5inch, 128x128)](https://www.sparkfun.com/products/15890))
+to the board and get started very qwiickly. It is worth mentioning that the
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:37]:
```suggestion
to the board and get started very qwiickly. It is worth mentioning that while
```
😂

File: content/posts/rust-and-riscv.md [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+advantage of the Qwiic connector is how easily and safely you can connect
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:38]:
```suggestion
the Qwiic connector makes it easy and safe to connect
```

File: content/posts/rust-and-riscv.md [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+components together but not the transfer speed of the data: connection through
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:39]:
```suggestion
components together, it is not fast: connection through
```

File: content/posts/rust-and-riscv.md [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+ *  [SparkFun RED-V Thing Plus](https://www.sparkfun.com/products/15799)
+ *  [Zio Qwiic OLED Display (1.5inch, 128x128)](https://www.sparkfun.com/products/15890)
+ *  [Qwiic Cable - 50mm](https://www.sparkfun.com/products/14426)
+
+**Note:** We don't need a USB to TTL serial cable this time.
+
+First step: making sure everything works
+----------------------------------------
+
+I assume that the screen is working because I already made it working in the
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:54]:
```suggestion
I assume that the screen is working because I already made it work in the
```

File: content/posts/rust-and-riscv.md [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+previous tutorial with the AVR. All we need to do now is making sure that the
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:55]:
```suggestion
previous tutorial with the AVR. All we need to do now is make sure that the
```

File: content/posts/rust-and-riscv.md [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+board itself is working by making a led blink. In this tutorial I'm not going
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:56]: Making a led blink is the hello world of embedded dev ;-)

File: content/posts/rust-and-riscv.md [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+    needs to keep running in the background. In one terminal you need to run:
+
+    ```text
+    sudo JLinkGDBServer -device FE310 -if JTAG -speed 4000 -port 3333 -nogui
+    ```
+
+    (You might not need `sudo` if you have set up user permissions to the
+    device).
+
+    The program should keeps waiting if it detected the device properly.
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:174]:
```suggestion
    The program should keep waiting if it detected the device properly.
```

File: content/posts/rust-and-riscv.md [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+    ```text
+    cargo run --example leds_blink
+    ```
+
+    The two programs will interact and load the new firmware. You should see
+    the led blinking.
+
+2.  By copying the hex file.
+
+    You first need to build the program:
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:187]:
```suggestion
    First, you need to build the program:
```

File: content/posts/rust-and-riscv.md [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+
+    Now you can generate the hex file from the binary using this command:
+
+    ```text
+    objcopy -S -j .text -j .rodata -O ihex \
+        ./target/riscv32imac-unknown-none-elf/debug/examples/leds_blink \
+        /media/your_device/leds_blink.hex
+    ```
+
+    Unmount. It should reboot automatically on the new firmware and you should
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:203]: Maybe make these steps an itemized list, so it's clear that you have to take these steps in the right order?

File: content/posts/rust-and-riscv.md [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+
+2.  By copying the hex file.
+
+    You first need to build the program:
+
+    ```text
+    cargo build --example leds_blink
+    ```
+
+    When you connect the device in USB you should see a new disk, mount it.
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:193]: Maybe put a reference to the paragraph higher up where you explain how you do this. (push the button etc.)

File: content/posts/rust-and-riscv.md [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+
+If you followed the previous tutorial
+["Rust, Arduino and Embedded Development as a Beginner: Part 1"]({{< ref "rust-and-arduino-part1" >}})
+this should be quite straightforward.
+
+You will need the animation frame we have generated in
+["Rust, Arduino and Embedded Development as a Beginner: Part 3"]({{< ref "rust-and-arduino-part3" >}}).
+Copy the RAW files to the `src/` directory.
+
+Make sure yo use nightly:
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:224]:
```suggestion
For the example, we are using nightly features, so make sure you tell rustup to use it:
```

File: content/posts/rust-and-riscv.md [hunk 19]
```diff
@@ -0,0 +1,408 @@
...
+    // Configure I2C
+    let sda = pin!(pins, i2c0_sda).into_iof0();
+    let scl = pin!(pins, i2c0_scl).into_iof0();
+    let mut i2c = I2c::new(p.I2C0, sda, scl, Speed::Fast, clocks);
+
+    // Get blue led
+    let sck = pin!(pins, spi0_sck);
+    let mut blue = sck.into_inverted_output();
+
+    let address = 0b0111100; // replace this by the address of your device
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:289]: How do you get this value?

File: content/posts/rust-and-riscv.md [hunk 20]
```diff
@@ -0,0 +1,408 @@
...
+Thus adding a lot of duplicated code. You can see the generated code of a macro
+by using
+[cargo-expand](https://crates.io/crates/cargo-expand). This is a very handy
+tool if you work with macros.
+
+My idea was that if I don't do a function call, it might be slightly faster.
+But it turns out that it was quite the opposite here because the FE310-G002 CPU
+has an instruction cache which can overflow if you have too many instructions
+in your loop, thus forcing the CPU to load the instructions from flash memory
+instead (which is slow). In this case it was a lot slower if `draw_frame` was a
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:407]:
```suggestion
instead (which is slow). In this case the macro version of `draw_frame` was a
```

File: content/posts/rust-and-riscv.md [hunk 21]
```diff
@@ -0,0 +1,408 @@
...
+        draw_frame(FRAME_14);
+        draw_frame(FRAME_15);
+    }
+}
+```
+
+Implementation notes
+--------------------
+
+I didn't use a macro this time but a closure instead. The difference between
```
Comment (wdullaer) [inline content/posts/rust-and-riscv.md:396]: my comment from last time wasn't entirely without merit :-D

Does the closure version also result in a smaller binary size?

**Review by wdullaer — COMMENTED** [review]
Nice post.
I have mostly just spelling and structure nitpicks.

The last section is hard to follow without having reread the previous blog posts.
I know you call this out, but either I'd repeat a bit more, or I'd spend a bit more time highlighting what has changed with the new board.

File: content/posts/rust-and-riscv.md (hunk 13 above)
Comment (cecton) [reply content/posts/rust-and-riscv.md:56]: :100:

File: content/posts/rust-and-riscv.md (hunk 18 above)
Comment (cecton) [reply content/posts/rust-and-riscv.md:224]: yo

File: content/posts/rust-and-riscv.md (hunk 19 above)
Comment (cecton) [reply content/posts/rust-and-riscv.md:289]: It was explain in the previous article. But maybe I can put a small reminder. This tutorial is not meant to be 100% beginner like the previous one

File: content/posts/rust-and-riscv.md (hunk 21 above)
Comment (cecton) [reply content/posts/rust-and-riscv.md:396]: Normally yes that is the whole point. But it's not because the binary is smaller that it is faster at all!

The reason why it was faster in this particular case is because the instruction cache is particularly small so there are a lot of cache miss which forces the MCU to fetch the next instructions from the flash.

On a PC the whole program is loaded in memory so if you can avoid a few jumps here and there in the code it might save some time.

File: content/posts/rust-and-riscv.md [hunk 22]
```diff
@@ -0,0 +1,408 @@
...
+by using
+[cargo-expand](https://crates.io/crates/cargo-expand). This is a very handy
+tool if you work with macros.
+
+My idea was that if I don't do a function call, it might be slightly faster.
+But it turns out that it was quite the opposite here because the FE310-G002 CPU
+has an instruction cache which can overflow if you have too many instructions
+in your loop, thus forcing the CPU to load the instructions from flash memory
+instead (which is slow). In this case the macro version of `draw_frame` was a
+macro.
```
Comment (cecton) [inline content/posts/rust-and-riscv.md:408]:
```suggestion
instead (which is slow). In this case the macro version of `draw_frame` was a
a lot slower.
```

File: content/posts/rust-and-riscv.md (hunk 17 above)
Comment (cecton) [reply content/posts/rust-and-riscv.md:193]: hmm? You don't press any button or anything here. The method 2 is totally unrelated to the method 1. You connect the thing in USB and the disk appear, it is as easy as that. Maybe I should make it clearer? I'm not sure what confused you
